### DISCUSSION TITLE
How to do pdf masking or text manipulation in pdf ?

### DISCUSSION THREAD

**Discussion by Laxmi530** [pymupdf/PyMuPDF#2412] [Q&A] [answered]
Hi,
I have a multi-page pdf wanted to mask certain fields in the pdf file, suppose like account number, phone number and name or wanted to replace other text in that place.

Can someone please guide me how to do this.
Thanking you in advance.

**Comment by JorjMcKie** [comment]
I presume you mean normal, standard text, when you say "field" - not PDF form fields?
The most important thing is how to identify the position of this items: the rectangle `rect` that wraps the respective text. Then use redaction annotations to remove all stuff inside the rectangle.
Like this:
```python
import fitz
doc = fitz.open("input.pdf")
for page in doc:
    # somehow locate the items to make a list of rectangle "rects"
    for rect in rects:
        page.add_redact_annot(rect)
    page.apply_redactions()
doc.ez_save("output.pdf")
```
So far the easy part.
The above will produce empty (white) rectangle where one of the "rect" are.
You can also use any other color or show some other text instead.
How do you plan to identify the single rectangles?

**Reply by Laxmi530** [reply]
Thanks @JorjMcKie for the response.
What actually I thought for the task is
i). First, I will extract the text from PDF.
ii). Will write regex to find the pattern like mobile number and email-id from the extracted text.
iii). Post finding the pattern will use `get_text("dict")["blocks"]` to get the text and respected bounding box.
iv). With the help of bounding box, we can replace other text.

Please guide me further.
Thanking you in advance.

**Reply by JorjMcKie** [reply]
I understand.
If you know the strings representing the mobile number and e-mail, you can simply search for the position on the page like this: `rect_list = page.search_for(mobile_number)`, similar for the e-mail. The response is a list of rectangles wrapping the searched item.
Use this to redact this information away as shown above.

**Reply by Laxmi530** [reply]
Thanks @JorjMcKie for the response. 
The above code is working but as you mentioned `You can also use any other color or show some other text instead.`  where I can pass the other text which I want to replace in the PDF.

Please guide me further.
Thanking you in advance.

**Reply by JorjMcKie** [reply]
Method `page.add_redact_annot()` adds a redaction. Please read the documentation for this. You can set a background color of the rectangle, and also specify the text that should replace the original.

Replacement text however is a tricky thing! It usually looks awkward or ugly, for one or more of the following reasons:
* Only the fonts Helvetica, Time-Roman, Courier and their variants are supported
* The new text will have a different length. If too long, multiple lines will be generated and the font size will be reduced in an efort to fit the new text in. If nothing of that helps, the rectangle remains empty.
* The insertion start point ("origin") can only be estimated, therefore no even, nice appearance might result.

For the above reasons, to achieve a nicer replacement appearance, it is often better to follow a step-wise approach:
1. extract detail properties of the to-be-deleted text (origin, font size, font, text color, etc.)
2. let the redaction process empty the rectangle
3. after `page.apply_redactions()`, insert the new text using a more appropriate font, the right origin, the right text color and font size

Step 1 can be done via `page.get_text("dict", clip=rect)` with the redaction rectangle. This delivers all you need.
Getting the right font for step 3 usually is tricky, because you often cannot use the original font from the PDF: in most cases that font is only a subset of the full font and thus has a limited number of supported characters. Your new text probably has other characters beyond those. So you will need the original full font - which can be a nightmare to find / download etc.

**Reply by Laxmi530** [reply]
Hai @JorjMcKie 
Thanks for your guidance.
I tried as you mentioned above but did not get the expected output. Below is the sample code and PDF file 
which I used.
```
import fitz, os
file = 'Test_Document.pdf' 
```
```
def rect_name_size_color(file, words):
    with fitz.open(file) as pdf_file:
        rects = [rect for item in words for page in pdf_file for rect in page.search_for(item)]
        fontname = [text['font'] for page in pdf_file for item in words for rect in page.search_for(item)
                    for text in [page.get_text("dict", clip=rect)['blocks'][0]['lines'][0]['spans'][0]]]
        fontsize = [text['size'] for page in pdf_file for item in words for rect in page.search_for(item)
                    for text in [page.get_text("dict", clip=rect)['blocks'][0]['lines'][0]['spans'][0]]]
        text_color = [text['color'] for page in pdf_file for item in words for rect in page.search_for(item)
                    for text in [page.get_text("dict", clip=rect)['blocks'][0]['lines'][0]['spans'][0]]]
    return rects, fontname, fontsize, text_color
```
```
old_words = ['1234567890', 'At glance 1']
new_words = ['123 456 789', 'At_glance_1']

file_name = os.path.basename(os.path.splitext(file)[0])
rects, font_name, font_size, textcolor = rect_name_size_color(file, old_words)
with fitz.open(file) as pdf_file:
    for page in pdf_file:
        for rect in rects:
            page.add_redact_annot(rect)
        page.apply_redactions()
        quad = None
        text = None
        fontname = None
        fontsize = None
        text_color = None
        for rect_1 in rects:
            quad = rect_1
            for word in new_words:
                text = word
                for fnt_name in font_name:
                    fontname = fnt_name
                    for fnt_size in font_size:
                        fontsize = fnt_size
                        for txt_clr in textcolor:
                            text_color = txt_clr
        page.add_redact_annot(quad=quad, text=text, fontname=fontname, fontsize=fontsize, text_color=text_color)
    pdf_file.ez_save(f"Edited_{file_name}.pdf")
```
[Test_Document.pdf](https://github.com/pymupdf/PyMuPDF/files/11544152/Test_Document.pdf)
[Edited_Test_Document.pdf](https://github.com/pymupdf/PyMuPDF/files/11544158/Edited_Test_Document.pdf)

Please guide me where is the mistake.
Thanking you in advance.

**Reply by JorjMcKie** [reply]
Once you have applied redactions, you will have emtpy rectangles where old text was deleted, right?
You then must **_insert_** the replacement text. Instead you have added redaction annots again ...🙁

**Reply by Laxmi530** [reply]
Thanks @JorjMcKie for the response.
Yes right also I followed the [Documentation](https://pymupdf.readthedocs.io/en/latest/page.html#Page.add_redact_annot:~:text=add_redact_annot(quad%2C%20text%3DNone%2C%20fontname%3DNone%2C%20fontsize%3D11%2C%20align%3DTEXT_ALIGN_LEFT%2C%20fill%3D(1%2C%201%2C%201)%2C%20text_color%3D(0%2C%200%2C%200)%2C%20cross_out%3DTrue)) but still I am not getting where is the mistake.

Please guide me the further steps.
Thanking you in advance.

**Reply by JorjMcKie** [reply, marked as answer]
Try this script:
[test-JM.zip](https://github.com/pymupdf/PyMuPDF/files/11548981/test-JM.zip)

**Reply by Laxmi530** [reply]
Thanks @JorjMcKie for the solutions. 
It is working.
